crates/model/src/lib.rs

```diff
@@ -1,3 +1,5 @@
+#![allow(clippy::needless_borrow)]
+
```

Comment:
This doesn't appear to be necessary, removing the borrows doesn't cause issues.

There are two cases here of:
```rust
let some_array: [u8; N];
// OLD
hex::encode(&some_array);
// NEW
hex::encode(some_array);
```

This works because `hex::encode` accepts an `AsRef<[u8]>`, which `[u8; N]` implements. Conceptually, they are slightly different - but they can be solved by:
```
hex::encode(some_array.as_slice());
```

That being said, the compiler should actually treat both cases the same anyway, and since `[u8; N]: Copy`, I wouldn't kill this lint for the whole module just because of this (and hide some actual incorrect uses of needless borrows).

Reply:
I would suggest:

<details><summary><code>`changes.patch`</code></summary>

```diff
diff --git a/crates/database/src/byte_array.rs b/crates/database/src/byte_array.rs
index 72f3b70b..79e48b97 100644
--- a/crates/database/src/byte_array.rs
+++ b/crates/database/src/byte_array.rs
@@ -58,7 +58,7 @@ impl<const N: usize> Decode<'_, Postgres> for ByteArray<N> {
 
 impl<const N: usize> Encode<'_, Postgres> for ByteArray<N> {
     fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> IsNull {
-        <&[u8] as Encode<Postgres>>::encode(&self.0, buf)
+        self.0.as_slice().encode(buf)
     }
 }
 
@@ -81,7 +81,7 @@ mod tests {
 
         let data: ByteArray<3> = ByteArray([1, 2, 3]);
         sqlx::query(&format!("INSERT INTO {} (bytes) VALUES ($1);", TABLE))
-            .bind(&data)
+            .bind(data)
             .execute(&db)
             .await
             .unwrap();
diff --git a/crates/database/src/lib.rs b/crates/database/src/lib.rs
index 1c5fdbd4..ea70dc97 100644
--- a/crates/database/src/lib.rs
+++ b/crates/database/src/lib.rs
@@ -1,5 +1,3 @@
-#![allow(clippy::needless_borrow)]
-
 pub mod auction;
 pub mod byte_array;
 pub mod ethflow_orders;
diff --git a/crates/database/src/onchain_broadcasted_orders.rs b/crates/database/src/onchain_broadcasted_orders.rs
index fd706ae5..138c465c 100644
--- a/crates/database/src/onchain_broadcasted_orders.rs
+++ b/crates/database/src/onchain_broadcasted_orders.rs
@@ -59,7 +59,7 @@ pub async fn insert_onchain_order(
     "#;
     sqlx::query(QUERY)
         .bind(event.order_uid)
-        .bind(&event.sender)
+        .bind(event.sender)
         .bind(index.block_number)
         .bind(index.log_index)
         .execute(ex)
diff --git a/crates/database/src/orders.rs b/crates/database/src/orders.rs
index 3075f99e..d0965860 100644
--- a/crates/database/src/orders.rs
+++ b/crates/database/src/orders.rs
@@ -153,9 +153,9 @@ INSERT INTO interactions (
 VALUES ($1, $2, $3, $4, $5)
     "#;
     sqlx::query(QUERY)
-        .bind(&order_uid)
-        .bind(&index)
-        .bind(&pre_interaction.target)
+        .bind(order_uid)
+        .bind(index)
+        .bind(pre_interaction.target)
         .bind(&pre_interaction.value)
         .bind(&pre_interaction.data)
         .execute(ex)
@@ -217,22 +217,22 @@ INSERT INTO orders (
 VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21)
     "#;
     sqlx::query(QUERY)
-        .bind(&order.uid)
-        .bind(&order.owner)
+        .bind(order.uid)
+        .bind(order.owner)
         .bind(order.creation_timestamp)
-        .bind(&order.sell_token)
-        .bind(&order.buy_token)
-        .bind(&order.receiver)
+        .bind(order.sell_token)
+        .bind(order.buy_token)
+        .bind(order.receiver)
         .bind(&order.sell_amount)
         .bind(&order.buy_amount)
         .bind(order.valid_to)
-        .bind(&order.app_data)
+        .bind(order.app_data)
         .bind(&order.fee_amount)
-        .bind(&order.kind)
+        .bind(order.kind)
         .bind(order.partially_fillable)
         .bind(order.signature.as_slice())
         .bind(order.signing_scheme)
-        .bind(&order.settlement_contract)
+        .bind(order.settlement_contract)
         .bind(order.sell_token_balance)
         .bind(order.buy_token_balance)
         .bind(&order.full_fee_amount)
@@ -308,7 +308,7 @@ buy_amount = $6
     "
     );
     sqlx::query(QUERY)
-        .bind(&quote.order_uid)
+        .bind(quote.order_uid)
         .bind(quote.gas_amount)
         .bind(quote.gas_price)
         .bind(quote.sell_token_price)
@@ -321,7 +321,7 @@ buy_amount = $6
 
 pub async fn insert_quote(ex: &mut PgConnection, quote: &Quote) -> Result<(), sqlx::Error> {
     sqlx::query(INSERT_ORDER_QUOTES_QUERY)
-        .bind(&quote.order_uid)
+        .bind(quote.order_uid)
         .bind(quote.gas_amount)
         .bind(quote.gas_price)
         .bind(quote.sell_token_price)
diff --git a/crates/database/src/quotes.rs b/crates/database/src/quotes.rs
index b0384e68..7ac72d57 100644
--- a/crates/database/src/quotes.rs
+++ b/crates/database/src/quotes.rs
@@ -7,7 +7,7 @@ use sqlx::{
 
 pub type QuoteId = i64;
 
-#[derive(Clone, Debug, Default, Eq, PartialEq, sqlx::Type)]
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, sqlx::Type)]
 #[sqlx(type_name = "QuoteKind")]
 #[sqlx(rename_all = "lowercase")]
 pub enum QuoteKind {
@@ -52,8 +52,8 @@ VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
 RETURNING id
     "#;
     let (id,) = sqlx::query_as(QUERY)
-        .bind(&quote.sell_token)
-        .bind(&quote.buy_token)
+        .bind(quote.sell_token)
+        .bind(quote.buy_token)
         .bind(&quote.sell_amount)
         .bind(&quote.buy_amount)
         .bind(quote.gas_amount)
@@ -61,7 +61,7 @@ RETURNING id
         .bind(quote.sell_token_price)
         .bind(quote.order_kind)
         .bind(quote.expiration_timestamp)
-        .bind(&quote.quote_kind)
+        .bind(quote.quote_kind)
         .fetch_one(ex)
         .await?;
     Ok(id)
@@ -111,14 +111,14 @@ ORDER BY gas_amount * gas_price * sell_token_price ASC
 LIMIT 1
     "#;
     sqlx::query_as(QUERY)
-        .bind(&params.sell_token)
-        .bind(&params.buy_token)
+        .bind(params.sell_token)
+        .bind(params.buy_token)
         .bind(&params.sell_amount_0)
         .bind(&params.sell_amount_1)
         .bind(&params.buy_amount)
         .bind(params.kind)
         .bind(params.expiration)
-        .bind(&params.quote_kind)
+        .bind(params.quote_kind)
         .fetch_optional(ex)
         .await
 }
@@ -403,7 +403,7 @@ mod tests {
             buy_amount: quote.buy_amount.clone(),
             kind: quote.order_kind,
             expiration: quote.expiration_timestamp,
-            quote_kind: quote.quote_kind.clone(),
+            quote_kind: quote.quote_kind,
         };
 
         assert_eq!(find(&mut db, &search_a).await.unwrap().unwrap(), quote,);
diff --git a/crates/model/src/lib.rs b/crates/model/src/lib.rs
index a603e6ac..cba97f2f 100644
--- a/crates/model/src/lib.rs
+++ b/crates/model/src/lib.rs
@@ -1,5 +1,3 @@
-#![allow(clippy::needless_borrow)]
-
 //! Contains models that are shared between the orderbook and the solver.
 
 pub mod app_id;
diff --git a/crates/model/src/order.rs b/crates/model/src/order.rs
index 9de9c8ed..cf35c472 100644
--- a/crates/model/src/order.rs
+++ b/crates/model/src/order.rs
@@ -501,7 +501,7 @@ impl Display for OrderUid {
         let mut bytes = [0u8; 2 + 56 * 2];
         bytes[..2].copy_from_slice(b"0x");
         // Unwrap because the length is always correct.
-        hex::encode_to_slice(&self.0, &mut bytes[2..]).unwrap();
+        hex::encode_to_slice(self.0, &mut bytes[2..]).unwrap();
         // Unwrap because the string is always valid utf8.
         let str = std::str::from_utf8(&bytes).unwrap();
         f.write_str(str)
diff --git a/crates/model/src/signature.rs b/crates/model/src/signature.rs
index edf2d09f..e60433e4 100644
--- a/crates/model/src/signature.rs
+++ b/crates/model/src/signature.rs
@@ -346,7 +346,7 @@ impl Serialize for EcdsaSignature {
         let mut bytes = [0u8; 2 + 65 * 2];
         bytes[..2].copy_from_slice(b"0x");
         // Can only fail if the buffer size does not match but we know it is correct.
-        hex::encode_to_slice(&self.to_bytes(), &mut bytes[2..]).unwrap();
+        hex::encode_to_slice(self.to_bytes(), &mut bytes[2..]).unwrap();
         // Hex encoding is always valid utf8.
         let str = std::str::from_utf8(&bytes).unwrap();
         serializer.serialize_str(str)

```

</details>

Reply:
The issue `https://github.com/rust-lang/rust-clippy/issues/9782` . This lint already existed before so we probably don't have false positives (but could in the future). It feels more correct to ignore the lint than change our code to abide by a bad lint .

Reply:
Ah OK - your example in the issue of it being a false positive makes perfect sense.